arangod/RocksDBEngine/RocksDBCollection.cpp

```diff
@@ -69,16 +73,499 @@
+  if (res.fail()) {
+    return res;
+  }
+  for (auto& idxPair : indicesMap) {
+    auto ridx = static_cast<arangodb::RocksDBIndex*>(idxPair.first.get());
+    res = ridx->setObjectIds(idxPair.second.first, idxPair.second.second);
+    if (res.fail()) {
+      return res;
+    }
+  }
+
+  return res;
+}
+
+arangodb::Result updateObjectIdsForCollection(rocksdb::DB& db,
+                                              arangodb::LogicalCollection& collection,
+                                              ::ObjectIdTransformer idFunc,
+                                              bool setUpgradedProperties) {
+  using arangodb::StaticStrings;
+  using arangodb::basics::VelocyPackHelper;
+
+  arangodb::Result res;
+  rocksdb::WriteOptions wo;
+  rocksdb::WriteBatch batch;
+
+  // methods need lock
+  arangodb::RocksDBKey key;  // read collection info from database
+  key.constructCollection(collection.vocbase().id(), collection.id());
+  rocksdb::PinnableSlice ps;
+  rocksdb::Status s =
+      db.Get(rocksdb::ReadOptions(),
+             arangodb::RocksDBColumnFamily::definitions(), key.string(), &ps);
+  if (!s.ok()) {
+    res.reset(arangodb::rocksutils::convertStatus(s));
+    return res;
+  }
+  arangodb::velocypack::Slice oldProps = arangodb::RocksDBValue::data(ps);
+
+  arangodb::velocypack::Builder builder;
+  std::pair<std::uint64_t, std::uint64_t> outputPair;
+  ::IndicesMap indicesMap;
+  {
+    arangodb::velocypack::ObjectBuilder collectionObjectGuard(&builder);
+    std::uint64_t objectId =
+        VelocyPackHelper::stringUInt64(oldProps, StaticStrings::ObjectId);
+    std::uint64_t tempObjectId =
+        VelocyPackHelper::stringUInt64(oldProps, StaticStrings::TempObjectId);
+    outputPair = idFunc(objectId, tempObjectId);
+    for (auto pair : arangodb::velocypack::ObjectIterator(oldProps)) {
+      ::handlePropertiesEntryForObjectIdUpdate(collection, builder, idFunc, setUpgradedProperties,
+                                               indicesMap, pair.key, pair.value);
+    }
+    builder.add(StaticStrings::ObjectId,
+                arangodb::velocypack::Value(std::to_string(outputPair.first)));
+    builder.add(StaticStrings::TempObjectId,
+                arangodb::velocypack::Value(std::to_string(outputPair.second)));
+  }
+  auto value = arangodb::RocksDBValue::Collection(builder.slice());
+  batch.Put(arangodb::RocksDBColumnFamily::definitions(), key.string(), value.string());
+  res = arangodb::rocksutils::convertStatus(db.Write(wo, &batch));
+  if (res.fail()) {
+    return res;
+  }
+  auto cleanup = arangodb::scopeGuard([oldProps, &db, &key]() -> void {
+    rocksdb::WriteOptions wo;
+    rocksdb::WriteBatch batch;
+    auto value = arangodb::RocksDBValue::Collection(oldProps);
+    batch.Put(arangodb::RocksDBColumnFamily::definitions(), key.string(), value.string());
+    db.Write(wo, &batch);
+  });
+
+  res = ::setObjectIdsForCollection(collection, outputPair, indicesMap);
```

Comment:
If there is a fatal error/system halt while we are in `setObjectIdsForCollection`, then the scopeGuard will not be executed anymore. Does this leave things in a broken state, or will we be able to recover from this state?

Reply:
We should be able to handle it in recovery. This `scopeGuard` is just to fix the state if we encounter an error and need to still function without restarting the server.